kep/tbd-peaks-power-aware-scheduling/README.md

```diff
@@ -0,0 +1,326 @@
+- PEAKS plugin requires the power model parameters for each cluster node configured with it. However, the power model (utilization vs. power consumption) parameters are subjected to change (e.g., increase in the number of CPUs on a node results in a change in the power model) over time. To address this, `configmap` is used as a mechanism to pass the updated power model parameters of any cluster node seamlessly.
+
+## Design Details
+
+### PEAKS workflow
+![PEAKS workflow](./figs/PEAKS_workflow.png)
+
+Here is a brief discription of some of the steps in the above workflow:
+- Create a power model for each cluster node: If a power model suitable for the workload to be scheduled already exists, then the same can be used to avoid creating a new power model. Thus it is an optional step.
+  - PEAKS plugin allows the cluster owners to bring their own power models that best represent both the node behavior and the workload characteristics.
+  - Training the power model is outside the scope of PEAKS plugin.
+  - PEAKS plugin only does power model inferencing.
+  - Here is an example that models the "CPU utilization vs. Power" relationship for a two node cluster.
+    - This model changes corresponding to a change either in the node configuration or in the workload characteristics.![Power-models-of-nodes](./figs/PowerModelsOfNodes.png)
+- Get the resource need for a pod to be scheduled: This step requires estimating the pod resource needs.
+  - Consider pod resource requirements (requests/limits) from its specification as an alternative.
+  - If there are one or more running pods with the same image (case of auto scaling), consider the average load or resource need across those pods.
+- Find the change in the utilization if the pod is to be scheduled on a node: This step requires the current utilization and the updated utilization after the pod placement on a node (so that the instantaneous power at both these utilizations can be found from the power model later)
+  - Current node utilization can be queried from Prometheus (with the metric source provider being a service like node-exporter or load-watcher).
+  - Add the change in utilization (corresponding to the pod resource requirement) to the current utilization to get the updated utilization
+- Return the change in the node's instantaneous power:
+  - The power consumption at the current utilization can be either found from the node metrics or by referring to the power model
+  - The power consumption at the updated utilization can be found by referring to the power model
+- Normalize plugin score: The score range can be provided by the user (otherwise consider the range [0, 100] as the default)
+- Choose a suitable weight for PEAKS plugin in kube-scheduler configuration: This is specific to the environment configuration.
```

Comment:
I may expect a concrete API design for the arguments of this plugin.

Reply:
Thanks a lot @Huang-Wei for this comment. We have updated the details as below.

=====
Normalize plugin score: The scores provided by different score plugins (which are in the range [framework.MinNodeScore, framework.MaxNodeScore]) are normalized to the range [0, 100] by the PEAKS plugin implementation of the ScoreExtensions interface function [NormalizeScore](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/interface.go#L602).

=====
Hope these details address your comment.

Reply:
Nvm. In the updated KEP I noticed the power model data is supposed to be passed in via ConfigMap. I assume that is the only input this plugin depends on right? In other words, there is no need to define further plugin-specific "parameter" .